File: rust/src/modules/serialpipe/serial_bridge_pipe.rs
#[cfg(target_os = "windows")]
use std::sync::atomic::{AtomicBool, Ordering};
#[cfg(target_os = "windows")]
use std::sync::Arc;
#[cfg(target_os = "windows")]
use std::thread::{self, JoinHandle};
#[cfg(target_os = "windows")]
use windows::core::{s, PCSTR};
#[cfg(target_os = "windows")]
use windows::Win32::Foundation::{
    CloseHandle, GetLastError, ERROR_IO_PENDING, ERROR_PIPE_CONNECTED, HANDLE,
    INVALID_HANDLE_VALUE, WAIT_OBJECT_0, LocalFree, HLOCAL,
};
#[cfg(target_os = "windows")]
use windows::Win32::Security::Authorization::{ConvertStringSecurityDescriptorToSecurityDescriptorA, SDDL_REVISION_1};
#[cfg(target_os = "windows")]
use windows::Win32::Security::{PSECURITY_DESCRIPTOR, SECURITY_ATTRIBUTES};
#[cfg(target_os = "windows")]
use windows::Win32::Storage::FileSystem::{ReadFile, FILE_FLAG_OVERLAPPED, PIPE_ACCESS_DUPLEX};
#[cfg(target_os = "windows")]
use windows::Win32::System::IO::{CancelIoEx, GetOverlappedResult, OVERLAPPED};
#[cfg(target_os = "windows")]
use windows::Win32::System::Pipes::{
    ConnectNamedPipe, CreateNamedPipeA, DisconnectNamedPipe,
    PIPE_READMODE_MESSAGE, PIPE_TYPE_MESSAGE, PIPE_WAIT,
};
#[cfg(target_os = "windows")]
use windows::Win32::System::Threading::{
    CreateEventA, ResetEvent, SetEvent, WaitForMultipleObjects, INFINITE,
};

#[cfg(target_os = "windows")]
pub type LogCallback = Box<dyn Fn(&str) + Send + Sync>;
#[cfg(target_os = "windows")]
pub type ForwardCallback = Box<dyn Fn(&str) -> bool + Send + Sync>;

#[cfg(target_os = "windows")]
pub struct SerialBridgePipe {
    log_callback: Arc<LogCallback>,
    forward_callback: Arc<ForwardCallback>,
    stop_event: isize,
    stop_flag: Arc<AtomicBool>,
    pipe_thread: Option<JoinHandle<()>>,
}

impl SerialBridgePipe {
    const PIPE_NAME: PCSTR = s!("\\\\.\\pipe\\corestation_serial_bridge");

    pub fn new(log_callback: LogCallback, forward_callback: ForwardCallback) -> Self {
        let stop_event = unsafe { CreateEventA(None, true, false, None).unwrap_or_default() };
        let log_arc = Arc::new(log_callback);

        Self {
            log_callback: log_arc,
            forward_callback: Arc::new(forward_callback),
            stop_event: stop_event.0 as isize,
            stop_flag: Arc::new(AtomicBool::new(false)),
            pipe_thread: None,
        }
    }

    pub fn start(&mut self) -> bool {
        let stop_event_handle = HANDLE(self.stop_event as *mut std::ffi::c_void);
        if stop_event_handle.is_invalid() {
            return false;
        }

        self.stop_flag.store(false, Ordering::SeqCst);
        unsafe { let _ = ResetEvent(stop_event_handle); }

        let stop_flag_clone = Arc::clone(&self.stop_flag);
        let log_clone = Arc::clone(&self.log_callback);
        let forward_clone = Arc::clone(&self.forward_callback);
        let stop_event_isize = self.stop_event;

        self.pipe_thread = Some(thread::spawn(move || {
            Self::pipe_thread_proc(stop_flag_clone, stop_event_isize, log_clone, forward_clone);
        }));

        true
    }

    pub fn stop(&mut self) {
        if self.stop_flag.swap(true, Ordering::SeqCst) { return; }
        
        let stop_event_handle = HANDLE(self.stop_event as *mut std::ffi::c_void);

        if self.stop_event != 0 {
            unsafe { let _ = SetEvent(stop_event_handle); }
        }

        if let Some(handle) = self.pipe_thread.take() {
            let _ = handle.join();
        }
    }

    fn pipe_thread_proc(
        stop_flag: Arc<AtomicBool>,
        stop_event_isize: isize,
        log_callback: Arc<LogCallback>,
        forward_callback: Arc<ForwardCallback>,
    ) {
        unsafe {
            let stop_event = HANDLE(stop_event_isize as *mut std::ffi::c_void);
            let mut sd: PSECURITY_DESCRIPTOR = PSECURITY_DESCRIPTOR::default();
            if ConvertStringSecurityDescriptorToSecurityDescriptorA(
                s!("D:(A;;GA;;;BA)"), SDDL_REVISION_1, &mut sd, None,
            ).is_err() { return; }

            let mut sa = SECURITY_ATTRIBUTES {
                nLength: std::mem::size_of::<SECURITY_ATTRIBUTES>() as u32,
                lpSecurityDescriptor: sd.0,
                bInheritHandle: false.into(),
            };

            while !stop_flag.load(Ordering::SeqCst) {
                let h_pipe = CreateNamedPipeA(
                    Self::PIPE_NAME,
                    PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED,
                    PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT,
                    1, 1024, 1024, 0, Some(&mut sa),
                ).unwrap_or_default();

                if h_pipe.is_invalid() {
                    thread::sleep(std::time::Duration::from_secs(1));
                    continue;
                }

                let mut ov_connect = OVERLAPPED::default();
                ov_connect.hEvent = CreateEventA(None, true, false, None).unwrap_or_default();
                
                let mut connected = ConnectNamedPipe(h_pipe, Some(&mut ov_connect as *mut _)).is_ok();
                let err = GetLastError();

                if !connected && err == ERROR_IO_PENDING {
                    let handles = [ov_connect.hEvent, stop_event];
                    let wait = WaitForMultipleObjects(&handles[..], false, INFINITE);
                    if wait.0 == windows::Win32::Foundation::WAIT_OBJECT_0.0 + 1 {
                        let _ = CancelIoEx(h_pipe, Some(&ov_connect as *const _));
                    } else {
                        connected = true;
                    }
                } else if !connected && err == ERROR_PIPE_CONNECTED {
                    connected = true;
                }

                if connected && !stop_flag.load(Ordering::SeqCst) {
                    loop {
                        let mut buffer = [0u8; 1024];
                        let mut bytes_read = 0;
                        let mut ov_read = OVERLAPPED::default();
                        ov_read.hEvent = CreateEventA(None, true, false, None).unwrap_or_default();

                        let read_ok = ReadFile(h_pipe, Some(&mut buffer), None, Some(&mut ov_read as *mut _)).is_ok();

                        if !read_ok {
                            let read_err = GetLastError();
                            if read_err == ERROR_IO_PENDING {
                                let handles = [ov_read.hEvent, stop_event];
                                let wait = WaitForMultipleObjects(&handles[..], false, INFINITE);
                                if wait.0 == windows::Win32::Foundation::WAIT_OBJECT_0.0 + 1 {
                                    let _ = CancelIoEx(h_pipe, Some(&ov_read as *const _));
                                    let _ = CloseHandle(ov_read.hEvent);
                                    break;
                                }
                                let _ = GetOverlappedResult(h_pipe, &ov_read, &mut bytes_read, false);
                            } else {
                                let _ = CloseHandle(ov_read.hEvent);
                                break;
                            }
                        } else {
                            let _ = GetOverlappedResult(h_pipe, &ov_read, &mut bytes_read, true);
                        }

                        let _ = CloseHandle(ov_read.hEvent);
                        if bytes_read == 0 { break; }

                        let payload = String::from_utf8_lossy(&buffer[..bytes_read as usize]).to_string();
                        let _ = (forward_callback)(&payload);
                    }
                }
                if !ov_connect.hEvent.is_invalid() { let _ = CloseHandle(ov_connect.hEvent); }
                let _ = DisconnectNamedPipe(h_pipe);
                let _ = CloseHandle(h_pipe);
            }
            let _ = LocalFree(Some(HLOCAL(sd.0)));
        }
    }
}

#[cfg(target_os = "windows")]
impl Drop for SerialBridgePipe {
    fn drop(&mut self) {
        self.stop();
        // Check if the isize is non-zero instead of calling .is_invalid()
        if self.stop_event != 0 {
            unsafe {
                // Cast the isize back to a HANDLE before closing
                let _ = CloseHandle(HANDLE(self.stop_event as *mut std::ffi::c_void)); 
            }
        }
    }
}

File: src/modules/serialpipe/SerialBridgePipe.cpp
#ifdef _WIN32
#include "SerialBridgePipe.h"
#include "WindowsIpcClientAuth.h"
#include "../../platform/WindowsPlatform.h"
#include <sddl.h>
#include <chrono>

const char* const SerialBridgePipe::PIPE_NAME = "\\\\.\\pipe\\corestation_serial_bridge";

SerialBridgePipe::SerialBridgePipe(WindowsPlatform* platform) : platform_(platform) {
    stopEvent_ = CreateEvent(NULL, TRUE, FALSE, NULL);
    if (!stopEvent_) {
        Log("Failed to create stop event");
    }
}

SerialBridgePipe::~SerialBridgePipe() {
    Stop();
    if (stopEvent_) {
        CloseHandle(stopEvent_);
        stopEvent_ = nullptr;
    }
}

void SerialBridgePipe::Log(const std::string& msg) {
    if (platform_) {
        platform_->logMessage("[SerialBridgePipe] " + msg);
    }
}

bool SerialBridgePipe::Start() {
    if (!stopEvent_) {
        Log("Cannot start: stop event unavailable");
        return false;
    }

    stop_ = false;
    ResetEvent(stopEvent_);
    pipeThread_ = std::thread([this]() { PipeThreadProc(); });
    return true;
}

void SerialBridgePipe::Stop() {
    if (stop_.exchange(true)) {
        return;
    }

    if (stopEvent_) {
        SetEvent(stopEvent_);
    }

    if (pipeThread_.joinable()) {
        pipeThread_.join();
    }
}

void SerialBridgePipe::PipeThreadProc() {
    // Restrict connections to members of BUILTIN\Administrators. Any other
    // caller's CreateFile on this pipe fails with ERROR_ACCESS_DENIED before
    // a single byte is exchanged.
    PSECURITY_DESCRIPTOR sd = nullptr;
    if (!ConvertStringSecurityDescriptorToSecurityDescriptorA(
            "D:(A;;GA;;;BA)", SDDL_REVISION_1, &sd, NULL)) {
        Log("Failed to build pipe security descriptor, refusing to start");
        return;
    }

    SECURITY_ATTRIBUTES sa = {};
    sa.nLength = sizeof(sa);
    sa.lpSecurityDescriptor = sd;
    sa.bInheritHandle = FALSE;

    Log("Serial bridge pipe listener started");

    while (!stop_.load()) {
        HANDLE hPipe = CreateNamedPipeA(
            PIPE_NAME,
            PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED,
            PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT,
            1,
            1024,
            1024,
            0,
            &sa);

        if (hPipe == INVALID_HANDLE_VALUE) {
            Log("Failed to create named pipe");
            std::this_thread::sleep_for(std::chrono::seconds(1));
            continue;
        }

        OVERLAPPED ovConnect = {};
        ovConnect.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
        BOOL connected = ConnectNamedPipe(hPipe, &ovConnect);
        DWORD err = GetLastError();

        if (!connected && err == ERROR_IO_PENDING) {
            HANDLE handles[2] = { ovConnect.hEvent, stopEvent_ };
            DWORD wait = WaitForMultipleObjects(2, handles, FALSE, INFINITE);
            if (wait == WAIT_OBJECT_0 + 1) {
                CancelIoEx(hPipe, &ovConnect);
            }
            else {
                connected = TRUE;
            }
        }
        else if (!connected && err == ERROR_PIPE_CONNECTED) {
            connected = TRUE;
        }

        if (connected && !stop_.load()) {
            // Authenticate before reading/forwarding a single byte -- see
            // WindowsIpcClientAuth.h and specs/001-secure-serial-ipc.
            // Administrator-only pipe ACL (above) is defense-in-depth only;
            // this check is the actual control (spec.md FR-001..FR-005).
            bool authenticated = IpcAuth::WindowsIsAuthenticated(
                hPipe, "[SerialBridgePipe] ",
                [this](const std::string& m) { Log(m); });

            if (!authenticated) {
                Log("WARNING: rejected unauthenticated IPC bridge connection");
                if (ovConnect.hEvent) {
                    CloseHandle(ovConnect.hEvent);
                }
                DisconnectNamedPipe(hPipe);
                CloseHandle(hPipe);
                continue;
            }

            for (;;) {
                char buffer[1024] = { 0 };
                DWORD bytesRead = 0;
                OVERLAPPED ovRead = {};
                ovRead.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);

                BOOL readOk = ReadFile(hPipe, buffer, sizeof(buffer), NULL, &ovRead);
                if (!readOk) {
                    DWORD readErr = GetLastError();
                    if (readErr == ERROR_IO_PENDING) {
                        HANDLE handles[2] = { ovRead.hEvent, stopEvent_ };
                        DWORD wait = WaitForMultipleObjects(2, handles, FALSE, INFINITE);
                        if (wait == WAIT_OBJECT_0 + 1) {
                            CancelIoEx(hPipe, &ovRead);
                            CloseHandle(ovRead.hEvent);
                            break;
                        }
                        GetOverlappedResult(hPipe, &ovRead, &bytesRead, FALSE);
                    }
                    else {
                        CloseHandle(ovRead.hEvent);
                        break;
                    }
                }
                else {
                    GetOverlappedResult(hPipe, &ovRead, &bytesRead, TRUE);
                }

                CloseHandle(ovRead.hEvent);

                if (bytesRead == 0) {
                    break;
                }

                std::string payload(buffer, bytesRead);
                if (platform_) {
                    if (!platform_->forwardSerialBridgeMessage(payload)) {
                        Log("ERROR: Failed to forward message to serial port");
                    }
                }
            }
        }

        if (ovConnect.hEvent) {
            CloseHandle(ovConnect.hEvent);
        }
        DisconnectNamedPipe(hPipe);
        CloseHandle(hPipe);
    }

    LocalFree(sd);
    Log("Serial bridge pipe listener stopped");
}

#endif // _WIN32
